PTFE Proxy and Service Discovery Process (#1272)
hashicorp/terraform-website: content/source/docs/enterprise/install/installer.html.md

`NOTE: Please take precautions on application outage when applying configuration change, i.e. wait for all runs to finish, prevent new runs to trigger`
4. Restart the Replicated services following [the instructions for your distribution](https://help.replicated.com/docs/native/customer-installations/installing-via-script/#restarting-replicated).

### Proxy and Service Discovery Process

> Workspaces in Terraform Enterprise can be executed in various methods according to selected workflows and for each selected workflow, it uses **remote backend** to perform [remote operation](https://www.terraform.io/docs/cloud/run/#remote-operations) unless **Execution Mode** of the workspace is set **Local**. Terraform CLI that is executed as part of the **remote operation** inside Terraform Enterprise follows the same standard process to perform [service discovery](https://www.terraform.io/docs/internals/remote-service-discovery.html) where Terraform Enterprise itself is the remote service for Terraform CLI, depends on your infrastructure setup, the fully qualified name of Terraform Enterprise may be required to include into **Proxy ByPass** configuration of the Terraform Enterprise application layer to allow Terraform CLI to communicate to Terraform Enterprise itself, to configure the proxy bypass address of Terraform Enterprise application, navigate to port **8800** under `/console/settings`:

Reviewer: ```suggestion
Unless the **Execution Mode** of a workspace is set **Local**, Terraform Enterprise executes [remote operations](https://www.terraform.io/docs/cloud/run/#remote-operations) and uses a **remote backend**. 

The Terraform CLI performs **remote operations** for Terraform Enterprise and uses [service discovery](https://www.terraform.io/docs/internals/remote-service-discovery.html) to find the Terraform Enterprise service itself. Depending on your infrastructure setup, you may need to include the fully qualified name of Terraform Enterprise in the **Proxy ByPass** configuration of Terraform Enterprise's application layer. 

Configure the proxy bypass address of Terraform Enterprise application on port **8800** on the path `/console/settings`:
```

Reviewer: Also please check this suggestion carefully. I'm not sure I 100% understood the text so I took my best guess.

Reviewer 2: I'm going to synthesize Judith's edits with what I think I'm reading in the original:

```suggestion
Unless the execution mode of a workspace is set to "local", Terraform Enterprise performs [remote operations](/docs/cloud/run/index.html#remote-operations), running Terraform in its own worker VMs. 

When running within Terraform Enterprise's worker VMs, Terraform uses [service discovery](/docs/internals/remote-service-discovery.html) to find the Terraform Enterprise service itself. Depending on your infrastructure setup, you may need to tell Terraform Enterprise not to access its own hostname via the proxy, so that Terraform can communicate with the Terraform Enterprise services.

To do this, add Terraform Enterprise's fully qualified hostname to the **Proxy Bypass** setting in Terraform Enterprise's dashboard. The proxy bypass setting can be found on port **8800** on the path `/console/settings`.
```

> Workspaces in Terraform Enterprise can be executed in various methods according to selected workflows and for each selected workflow, it uses **remote backend** to perform [remote operation](https://www.terraform.io/docs/cloud/run/#remote-operations) unless **Execution Mode** of the workspace is set **Local**. Terraform CLI that is executed as part of the **remote operation** inside Terraform Enterprise follows the same standard process to perform [service discovery](https://www.terraform.io/docs/internals/remote-service-discovery.html) where Terraform Enterprise itself is the remote service for Terraform CLI, depends on your infrastructure setup, the fully qualified name of Terraform Enterprise may be required to include into **Proxy ByPass** configuration of the Terraform Enterprise application layer to allow Terraform CLI to communicate to Terraform Enterprise itself, to configure the proxy bypass address of Terraform Enterprise application, navigate to port **8800** under `/console/settings`:

Reviewer: The screenshot doesn't match the path `8800/console/settings`. It looks like `8800/settings` instead. Are we sure this is the right path or screenshot?

![Terraform Enterprise Proxy Bypass](./assets/tfe-proxy-bypass.png)

> Once configuration has been saved, please proceed to restart the application.

Reviewer: Does the user need to save the configuration or does TFE do it automatically?

Author: User will need to save the configuration, that would be the same to all other configurations, I think.

> To automate this step, the equivalent setting [extra_no_proxy](./automating-the-installer.html#available-settings) can be used.

Reviewer: ```suggestion
You can use the equivalent setting, [extra_no_proxy](./automating-the-installer.html#available-settings), to automate this step.
```

> Additionally, when using CA that is not recognized by Terraform Enterprise **worker image** to configure TLS of the Terraform Enterprise, it is suggested to add the CA into [CA Custom Bundle](certificate-authority-ca-bundle) configuration so that Terraform Enterprise will make the CA available in the worker container that execute Terraform CLI to be able to perform TLS handshake with Terraform Enterprise during service discovery or adding the CA into certificate file of the [custom image](alternative-terraform-worker-image) which is located at `/etc/ssl/certs/ca-certificates.crt` when non-default work image is configured.

Reviewer: ```suggestion
The Terraform CLI performs a TLS handshake with Terraform Enterprise during service discovery, and will need access to the Certificate Authority that Terraform Enterprise uses. 

Either add the CA into the [CA Custom Bundle](certificate-authority-ca-bundle) configuration so that Terraform Enterprise will make the CA available in the worker container, or add the CA directly into certificate file of the [custom image](alternative-terraform-worker-image) located at `/etc/ssl/certs/ca-certificates.crt` when a non-default work image is configured.
```

Reviewer: Please review this carefully as well. 

The links in the second paragraph seem broken. I wasn't able to resolve them on the live site. Where should they go?

Reviewer 2: It looks like the links are meant to be #anchors. 

```suggestion
The Terraform CLI performs a TLS handshake with Terraform Enterprise during service discovery, and will need access to the Certificate Authority that Terraform Enterprise uses. 

Either add the CA to the [CA Custom Bundle](#certificate-authority-ca-bundle) configuration so that Terraform Enterprise will make the CA available in the worker container, or, if a [custom worker image](#alternative-terraform-worker-image) is configured, add the CA directly to its certificate file (located at `/etc/ssl/certs/ca-certificates.crt`).
```
